lib/provisioning/ProvisionDispatcher.js

```diff
@@ -0,0 +1,355 @@
+'use strict'; // eslint-disable-line
+
+const async = require('async');
+const crypto = require('crypto');
+const path = require('path');
+const zookeeper = require('node-zookeeper-client');
+
+const Logger = require('werelogs').Logger;
+
+function _genRandomHexChars(len) {
+    return crypto.randomBytes(Math.ceil(len / 2))
+        .toString('hex') // convert to hexadecimal format
+        .slice(0, len);  // return required number of characters
+}
+
+class ProvisionDispatcher {
+
+    /**
+     * Create a new provision dispatcher object
+     *
+     * @constructor
+     * @param {object} zkConfig - zookeeper config object
+
+     * @param {string} zkConfig.endpoint - zookeeper endpoint string
+     * where provisions are managed
+     * (e.g. "localhost:2181/shared-tasks")
+     * @param {object} logConfig - logging configuration
+     * @param {string} logConfig.logLevel - logging level
+     * @param {string} logConfig.dumpLevel - dump level
+     */
+    constructor(zkConfig, logConfig) {
+        this.zkEndpoint = zkConfig.endpoint;
+        this.client = zookeeper.createClient(zkConfig.endpoint);
+        this.client.connect();
+        this.connectWaitList = [];
+        this.client.once('connected', () => {
```

Comment:
In the handler of `connected` event, we set `connectWaitList` to `null`, so that next calls to `_waitConnected` return immediately. AFAIK, if disconnects occur next, it's all up to the client zookeeper lib to attempt to reconnect automatically, transparently to the user. With the use of `once` handler, we'll just ignore next instances of `connected` event.

If we lose the zookeeper session though, we would have to restart the queue populator since we don't attempt to reestablish a new session when such event occurs. I think it's ok for now as the queue populator is local to a zookeeper node but it may have to be done in the future.